File: Data Structures & Algorithms/three-integer-sum/submission-3.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        std::vector<std::vector<int>> result;
        std::sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] > 0) break; // Break if current/leftmost value is > 0 - no more triplets
            if (i > 0 && nums[i] == nums[i-1]) continue; // Continue if current value is same as previous

            int left = i + 1;
            int right = nums.size() - 1;

            while (left < right)
            {
                int sum = nums[i] + nums[left] + nums[right];

                if (sum > 0) right--;
                else if (sum < 0) left++;
                else
                {
                    result.push_back({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left-1])
                    {
                        left++;
                    }
                }
            }


        }

        return result;
    }
};

File: Data Structures & Algorithms/permutation-string/submission-3.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False

        seenCount = 0
        left, right = 0, 0

        s1Chars = {}

        for c in s1:
            s1Chars[c] = s1Chars.get(c,0) + 1

        while right < len(s2) and s2[right] not in s1Chars:
            right += 1
            left += 1
            
        while right < len(s2):
            c = s2[right]
            if c not in s1Chars:
                while left < right:
                    if s2[left] in s1Chars:
                        s1Chars[s2[left]] += 1
                        seenCount -= 1
                    left += 1
            elif s1Chars[c] == 0:
                while s1Chars[c] == 0:
                    s1Chars[s2[left]] += 1
                    seenCount -= 1
                    left += 1
                    
            if c in s1Chars:
                s1Chars[c] -= 1
                seenCount += 1

            print(f"{c}, {left}, {right}, {s1Chars}, {seenCount}")
            if seenCount == len(s1):
                return True

            right += 1
        return False
